plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,545 @@
+	};
+
+	mbim_device_new (mbim_device_file, NULL, fu_mbim_qdu_updater_mbim_device_new_ready, &ctx);
+	g_main_loop_run (mainloop);
+
+	/* either we have all device or otherwise error is set */
+	if (ctx.mbim_device != NULL) {
+		g_warn_if_fail (!ctx.error);
+		self->mbim_device = ctx.mbim_device;
+		/* success */
+		return TRUE;
+	}
+
+	g_warn_if_fail (ctx.error != NULL);
+	g_warn_if_fail (ctx.mbim_device == NULL);
+	g_propagate_error (error, ctx.error);
+	return FALSE;
+}
+
+typedef struct {
+	GMainLoop	*mainloop;
+	MbimDevice	*mbim_device;
+	guint		 timeout_id;
+	GError		*error;
+} CloseContext;
+
+static void
+fu_mbim_qdu_updater_mbim_device_close_ready (GObject *mbim_device, GAsyncResult *res, gpointer user_data)
+{
+	CloseContext *ctx = (CloseContext *) user_data;
+
+	/* ignore errors when closing */
+	mbim_device_close_finish (MBIM_DEVICE (mbim_device), res, (ctx->error == NULL) ? &ctx->error : NULL);
+	g_clear_object (&ctx->mbim_device);
+	g_main_loop_quit (ctx->mainloop);
+}
+
+static gboolean
+fu_mbim_qdu_updater_reprobe_timeout (gpointer user_data)
+{
+	CloseContext *ctx = user_data;
+
+	ctx->timeout_id = 0;
+	g_main_loop_quit (ctx->mainloop);
+
+	return G_SOURCE_REMOVE;
+}
+
+gboolean
+fu_mbim_qdu_updater_close (FuMbimQduUpdater *self, GError **error)
+{
+	g_autoptr(GMainLoop) mainloop = g_main_loop_new (NULL, FALSE);
+	g_autofree gchar *device_sysfs_path = NULL;
+	CloseContext ctx = {
+		.mainloop = mainloop,
+		.mbim_device = g_steal_pointer (&self->mbim_device),
+		.timeout_id = 0,
+	};
+
+	mbim_device_close (ctx.mbim_device, 5, NULL, fu_mbim_qdu_updater_mbim_device_close_ready, &ctx);
+	g_main_loop_run (mainloop);
+
+	/* we should always have both device cleared, and optionally error set */
+
+	g_warn_if_fail (ctx.mbim_device == NULL);
+
+	if (ctx.error != NULL) {
+		g_propagate_error (error, ctx.error);
+		return FALSE;
+	}
+ 
+	fu_mm_utils_get_port_info (self->mbim_port, &device_sysfs_path, NULL, NULL);
+	/* device will auto reboot after update */
+	/* pcie rescan is required */
+	/* mbim-proxy stuck due to mbim port suddenly lost */
+	if (device_sysfs_path) {
+		g_auto(GStrv) split = NULL;
+		split = g_strsplit (device_sysfs_path, "/", -1);
+		/* /sys/devices/pci0000:00/0000:00:1d.3/0000:02:00.0 */
+		if (g_strv_length (split) == 6) {
+			const gchar *remove_path = g_strdup_printf ("/sys/bus/pci/devices/%s/%s/remove", split[4], split[5]);
+			const gchar *rescan_path = "/sys/bus/pci/rescan";
+			gchar value;
+			FILE *f;
+			g_auto(GStrv) argc = NULL;
+			GError *tmp_error = NULL;
+			gint exit_stat = 0;
+
+			if (!(f = fopen (remove_path, "w"))) {
+				g_debug ("fail to open %s", remove_path);
+			}
+
+			if (f) {
+				value = '1';
+				if (fwrite (&value, 1, 1, f) != 1) {
+					g_debug ("fail to write %s", remove_path);
+				}
+			}
+
+			if (f)
+				fclose(f);
+
+			argc = g_new0 (gchar *, 4);
+			argc[3] = NULL;
+			argc[0] = g_strdup ("killall");
+			argc[1] = g_strdup ("-9");
+			argc[2] = g_strdup ("mbim-proxy");
```

Comment:
oh my god no

Reply:
Just think: the modem you're upgrading and is inhibited in MM and you can use it in fwupd, but there may be 30 additional MBIM modems also using the mbim-proxy, and you just broke the communication for all of them.

Reply:
Yes, you are right, the mbim-proxy problem need to be fixed rather than be killed, I will not kill it here.

The remove and rescan operations used because when module reboot after download firmware, the driver will not initialize successfully, maybe it is our hardware defects. I will get rid of remove and rescan here, what do you think? 
Or could you give me some suggestion?  Thank you very much.

Reply:
I would remove this logic from the patch, and try to see how to solve it, maybe it's a kernel driver issue that is better fixed in the kernel instead of via workarounds in userspace?

Reply:
> Yes, you are right, the mbim-proxy problem need to be fixed rather than be killed, I will not kill it here.

If there is any problem with the mbim-proxy, please report it as an issue to libmbim including verbose logs of the proxy itself, see https://modemmanager.org/docs/modemmanager/debugging/#qmi-proxy-or-mbim-proxy-debug-logs. I'd love to help fixing that problem!

Reply:
> If there is any problem with the mbim-proxy, please report it as an issue to libmbim including verbose logs of the proxy itself,

I have report the issue to libmbim including verbose logs of the mbim-proxy, https://gitlab.freedesktop.org/mobile-broadband/libmbim/-/issues/17. Thank you for helping me fix this problem.